### PR TITLE
Array support null value

### PR DESCRIPTION
**Pull request by voyager-dw**
support null value for array

### CODE REVIEW

**Comment by shz117** [comment]
can you document list type data format for both live and archive batches in the [wiki](https://github.com/uber/aresdb/wiki/Vector-Store)? 
It should help review and external OSS community to understand.

File: memstore/list/archive_vector_party.go [hunk 1]
```diff
@@ -139,16 +140,19 @@ func (vp *ArchiveVectorParty) GetDataValueByRow(row int) common.DataValue {
...
-		vp.SetOffsetLength(row, unsafe.Pointer(&zero), unsafe.Pointer(&zero))
+		vp.SetOffsetLength(row, unsafe.Pointer(&zero), unsafe.Pointer(&zero), false)
 		if row >= vp.lengthFilled {
 			vp.lengthFilled++
 		}
 		return
 	}
-	newLen := *(*uint32)(val)
+	var newLen uint32
+	if uintptr(val) != 0 {
```
Comment (jshencode) [inline memstore/list/archive_vector_party.go]: what does it mean when valid is true and val ptr is 0

File: memstore/common/data_value.go [hunk 2]
```diff
@@ -32,6 +33,13 @@ var NullDataValue = DataValue{}
 // SizeOfGeoPoint is the size of GeoPointGo in memory
 const SizeOfGeoPoint = unsafe.Sizeof(GeoPointGo{})
 
+// ZeroLengthArrayFlag is the value to represent 0 length array value
+// used in offset part of OffsetLength vector
+// when length part is non-zero value, the offset is real offset in memory
+// when length is 0 and offset is 0, the array value is invalid (by default)
+// when length is 0 and offset is ZeroLengthArrayFlag, this is a 0 length array value
```
Comment (jshencode) [inline memstore/common/data_value.go]: seems like we just need offset to be nonzero right for this case?

File: memstore/list/archive_vector_party.go (hunk 1 above)
Comment (jshencode) [reply memstore/list/archive_vector_party.go]: nvm, got it, but actually val != nil should be enough

File: memstore/list/live_vector_party.go [hunk 3]
```diff
@@ -287,17 +289,19 @@ func (vp *LiveVectorParty) SetValue(row int, val unsafe.Pointer, valid bool) {
 
 	var newLen int
 	if valid {
-		newLen = int(*(*uint32)(val))
+		if uintptr(val) != 0 {
```
Comment (jshencode) [inline memstore/list/live_vector_party.go]: same, val != nil should be better

File: memstore/list/live_vector_party.go [hunk 4]
```diff
@@ -337,9 +341,11 @@ func (vp *LiveVectorParty) GetValue(row int) (val unsafe.Pointer, validity bool)
 	vp.RLock()
 	defer vp.RUnlock()
 
-	offset, length := vp.GetOffsetLength(row)
-	if offset == 0 && length == 0 {
+	offset, length, valid := vp.GetOffsetLength(row)
+	if !valid {
 		return unsafe.Pointer(uintptr(0)), false
+	} else if length == 0 {
+		return unsafe.Pointer(uintptr(0)), true
```
Comment (jshencode) [inline memstore/list/live_vector_party.go]: i think we can just return nil, true

File: memstore/list/vector_party.go [hunk 5]
```diff
@@ -38,19 +38,31 @@ type baseVectorParty struct {
...
 	length = *(*uint32)(vp.offsets.GetValue(2*row + 1))
+	valid = !(length == 0 && offset == 0)
 	return
 }
 
 // SetOffsetLength update offset/length for nth fow
-func (vp *baseVectorParty) SetOffsetLength(row int, offset, length unsafe.Pointer) {
-	vp.offsets.SetValue(2*row, offset)
-	vp.offsets.SetValue(2*row+1, length)
+func (vp *baseVectorParty) SetOffsetLength(row int, offset, length unsafe.Pointer, valid bool) {
```
Comment (jshencode) [inline memstore/list/vector_party.go]: seems we can get validity from offset and length already, no need for the extra bool argument

File: query/iterator.hpp [hunk 6]
```diff
@@ -421,13 +421,13 @@ class ArrayVectorPartyIterator
...
     if (length == 0) {
-        return thrust::make_tuple(reinterpret_cast<Value*>(NULL), true);
-    } else if (length == UINT32_MAX) {
+      if (offset == 0) {
         return thrust::make_tuple(reinterpret_cast<Value*>(NULL), false);
-    } else {
-        return thrust::make_tuple(
-                reinterpret_cast<Value*>(valuePtr + offset), true);
+      }
+      return thrust::make_tuple(reinterpret_cast<Value*>(NULL), true);
```
Comment (jshencode) [inline query/iterator.hpp]: can be simplified to return thrust::make_tuple(reinterpret_cast<Value*>(NULL), offset == 0);

File: memstore/common/data_value.go (hunk 2 above)
Comment (voyager-dw) [reply memstore/common/data_value.go]: offset is not enouth, there is real valid cases when offset is 0

Comment (jshencode) [reply memstore/common/data_value.go]: no i mean when length is 0 we just need to know whether offset is zero or not to determine invalid or not, right?

Comment (voyager-dw) [reply memstore/common/data_value.go]: that is correct

File: memstore/common/data_value.go [hunk 7]
```diff
@@ -715,11 +723,17 @@ type ArrayValueReader struct {
 
 // NewArrayValueReader is to create ArrayValueReader to read from upsertbatch, which includes the item number
 func NewArrayValueReader(dataType DataType, value unsafe.Pointer) *ArrayValueReader {
-	length := *((*uint32)(value))
+	if value == nil || uintptr(value) == 0 {
```
Comment (jshencode) [inline memstore/common/data_value.go]: is value == nil enough?

Comment (voyager-dw) [reply memstore/common/data_value.go]: good catch, cleaned up

**Pull request merged**